boa_engine/src/optimizer/cfg.rs

```diff
@@ -0,0 +1,640 @@
+            | Opcode::Reserved32
+            | Opcode::Reserved33
+            | Opcode::Reserved34
+            | Opcode::Reserved35
+            | Opcode::Reserved36
+            | Opcode::Reserved37
+            | Opcode::Reserved38
+            | Opcode::Reserved39
+            | Opcode::Reserved40
+            | Opcode::Reserved41
+            | Opcode::Reserved42
+            | Opcode::Reserved43
+            | Opcode::Reserved44
+            | Opcode::Reserved45
+            | Opcode::Reserved46
+            | Opcode::Reserved47
+            | Opcode::Reserved48
+            | Opcode::Reserved49
+            | Opcode::Reserved50
+            | Opcode::Reserved51
+            | Opcode::Reserved52 => unreachable!("Reserved opcodes are unrechable"),
+        }
+
+        let end_operand_byte = self.pc;
+
+        Some(BytecodeIteratorResult {
+            current_opcode_pc,
+            next_opcode_pc: self.pc,
+            opcode,
+            operands: &self.bytecode[start_operand_byte..end_operand_byte],
+        })
+    }
+}
+
+impl FusedIterator for BytecodeIterator<'_> {}
+
+/// TODO: doc
+#[derive(Clone, Copy)]
+pub enum Terminator {
+    /// TODO: doc
+    None,
+    /// TODO: doc
+    JumpUnconditional(u32),
+    /// TODO: doc
+    JumpConditional(Opcode, u32),
+}
+
+impl Debug for Terminator {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Terminator::None => write!(f, "None"),
+            Terminator::JumpUnconditional(target) => write!(f, "Jump B{target}"),
+            Terminator::JumpConditional(opcode, target) => {
+                write!(f, "{} B{target}", opcode.as_str())
+            }
+        }
+    }
+}
+
+/// TODO: doc
+pub struct BasicBlock {
+    bytecode: Vec<u8>,
+    terminator: Terminator,
+}
+
+impl Debug for BasicBlock {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // writeln!(f, "BasicBlock")?;
+        for result in BytecodeIterator::new(&self.bytecode) {
+            writeln!(
+                f,
+                "    {:06}      {}",
+                result.current_opcode_pc,
+                result.opcode.as_str()
+            )?;
+        }
+        writeln!(f, "    Terminator: {:?}", self.terminator)?;
+
+        Ok(())
+    }
+}
+
+/// TODO: doc
+pub struct ControlFlowGraph {
+    basic_blocks: Vec<BasicBlock>,
+    references: Vec<Vec<u32>>,
+}
+
+impl Debug for ControlFlowGraph {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(f, "BasicBlocks:")?;
+
+        for (i, basic_block) in self.basic_blocks.iter().enumerate() {
+            write!(f, "  B{i}: -- referenced by ")?;
+            for r in &self.references[i] {
+                write!(f, "B{r}, ")?;
+            }
+            writeln!(f, "")?;
+
+            writeln!(f, "{basic_block:#?}")?;
+        }
+
+        Ok(())
+    }
+}
+
+fn is_jump_kind_opcode(opcode: Opcode) -> bool {
+    match opcode {
```

Comment:
This can probably be exchanged by a `matches!()` call, which could make it `const`